vsgenstubs4/_vapoursynth.part.pyi

```diff
@@ -516,236 +843,327 @@ class VideoFormat:
+    def __init__(self) -> NoReturn: ...  # type: ignore[misc]
 
-    def copy(self) -> typing.Dict[str, _FramePropsValue]: ...
+    def setdefault(
+        self, key: str, default: _VapourSynthMapValue = 0
+    ) -> _VapourSynthMapValue: ...
 
-    def __getattr__(self, name: str) -> _FramePropsValue: ...
-    def __setattr__(self, name: str, value: _FramePropsValue) -> None: ...
+    def copy(self) -> 'FrameProps': ...
+
+    def __setattr__(self, name: str, value: _VapourSynthMapValue) -> None: ...
+
+    def __getattr__(self, name: str) -> _VapourSynthMapValue: ...
+
+    def __delattr__(self, name: str) -> None: ...
+
+    def __setitem__(self, name: str, value: _VapourSynthMapValue) -> None: ...
+
+    def __getitem__(self, name: str) -> _VapourSynthMapValue: ...
 
-    # mypy lo vult.
-    # In all seriousness, why do I need to manually define them in a typestub?
     def __delitem__(self, name: str) -> None: ...
-    def __setitem__(self, name: str, value: _FramePropsValue) -> None: ...
-    def __getitem__(self, name: str) -> _FramePropsValue: ...
-    def __iter__(self) -> typing.Iterator[str]: ...
+
+    def __iter__(self) -> Iterator[str]: ...
+
     def __len__(self) -> int: ...
 
 
-class _RawFrame:
+class RawFrame:
+    def __init__(self) -> NoReturn: ...  # type: ignore[misc]
+
     @property
-    def readonly(self) -> bool: ...
+    def closed(self) -> bool: ...
+
+    def close(self) -> None: ...
+
+    def copy(self: 'SelfFrame') -> 'SelfFrame': ...
 
     @property
     def props(self) -> FrameProps: ...
 
-    def get_read_ptr(self, plane: int) -> ctypes.c_void_p: ...
-    def get_write_ptr(self, plane: int) -> ctypes.c_void_p: ...
+    @props.setter
+    def props(self, new_props: MappingProxyType[str, _VapourSynthMapValue]) -> None: ...
+
+    def get_write_ptr(self, plane: int) -> c_void_p: ...
+
+    def get_read_ptr(self, plane: int) -> c_void_p: ...
+
     def get_stride(self, plane: int) -> int: ...
 
     @property
-    def closed(self) -> bool: ...
+    def readonly(self) -> bool: ...
 
-    def close(self) -> None: ...
-    def __enter__(self) -> '_RawFrame': ...
-    def __exit__(self, ty: typing.Optional[typing.Type[BaseException]], tv: typing.Optional[BaseException], tb: typing.Optional[types.TracebackType]) -> None: ...
+    def __enter__(self: 'SelfFrame') -> 'SelfFrame': ...
 
+    def __exit__(
+        self, exc_type: Union[Type[BaseException], None],
+        exc_value: Union[BaseException, None],
+        traceback: Union[TracebackType, None], /,
+    ) -> Union[bool, None]: ...
 
-class VideoFrame(_RawFrame):
-    height: int
-    width: int
+    def __getitem__(self, index: int) -> memoryview: ...
+
+    def __len__(self) -> int: ...
+
+
+SelfFrame = TypeVar('SelfFrame', bound=RawFrame)
+
+
+class VideoFrame(RawFrame):
     format: VideoFormat
+    width: int
+    height: int
 
-    def copy(self) -> 'VideoFrame': ...
+    def _writelines(self, write: Callable[[bytes], None]) -> None: ...
```

Comment:
While this function is useful, it is still an internal function.
Please provide a PR on vapoursynth/vapoursynth first renaming this function, if you like to keep it in this file.

Reply:
Opened [vapoursynth/#901](https://github.com/vapoursynth/vapoursynth/pull/901)